File: src/com/jepl/lang/Values.java
package com.jepl.lang;

import com.jepl.lang.libraries.Function;
import com.jepl.lang.libraries.Library;

import java.util.HashMap;

public class Values {
    public static HashMap<String, String> funs = new HashMap<>();
    public static HashMap<String, Object> vars = new HashMap<>();
    public static HashMap<String, HashMap<String, Function>> libs = new HashMap<>();

    public static void addFun(String name, String body){
        if(!funs.containsKey(name)) {
            funs.put(name, body);
        }else {
            System.err.println("Error! Function "+name+" already defined.");
        }
    }

    public static String getFun(String name){
        if(funs.containsKey(name)){
            return funs.get(name);
        }
        return String.valueOf(0);
    }

    public static void addVar(String name, Object value) {
        vars.put(name, value);
    }

    public static void addLib(String name){
        Library lib;
        try {
            lib = (Library) Class.forName("com.jepl.lang.libraries."+name+"."+name).newInstance();
        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
            throw new JEPLException(new RuntimeException(e));
        }
        libs.put(name, lib.invoke());
    }

    public static String getVar(String name){
        if(vars.containsKey(name)){
            return vars.get(name).toString();
        }else {
            throw new JEPLException(new Throwable("Error! Variable "+name+" is not exists."));
        }
    }
}

File: src/com/jepl/lang/Values.kt
package com.jepl.lang

import com.google.gson.JsonParser
import com.jepl.lang.libraries.Function
import com.jepl.lang.libraries.Library

class Values {
    companion object {
        var funs: HashMap<String, String> = HashMap()
        var vars: HashMap<String, Any> = HashMap()
        var libs: HashMap<String, HashMap<String, Function>> = HashMap()

        fun addFun(name: String, body: String){
            if (!funs.containsKey(name)) {
                funs[name] = body
            } else {
                throw JEPLException(RuntimeException("Error [block ${Interpreter.block+1}] - function $name redefine"))
            }
        }

        fun getFun(name: String) : String? {
            return funs[name]
        }

        fun addVar(name: String, value: Any) {
            vars[name] = value
        }

        fun addLib(name: String) {
            val lib: Library
            try {
                lib = Class.forName("com.jepl.lang.libraries.$name.$name").newInstance() as Library
            }catch (e: NullPointerException) {
                throw JEPLException(e, JsonParser.parseString("{\"name\": \"?\", \"args\": []}").asJsonObject)
            }catch (e: ClassNotFoundException) {
                throw JEPLException(e, JsonParser.parseString("{\"name\": \"?\", \"args\": []}").asJsonObject)
            }
            libs[name] = lib.invoke()
        }
    }
}